### PR TITLE
Adding device  authorization oauth2 flow 

### PR DESCRIPTION
**Pull request by pmahindrakar-oss**
Signed-off-by: Prafulla Mahindrakar <prafulla.mahindrakar@gmail.com>

# TL;DR
Implements device authorization flow for flyte clients such as flytectl and also fixes token refresh bug where it won't refresh the token 

* Introduces new authType enum value DeviceFlow
```
flytectl get projects --logger.level=6  --admin.authType DeviceFlow
```
* Augments the FlyteClientConf to include the device authorization endpoint url which is fetched from the metadata endpoint of admin and externally it gets directly from the authorization servers well known configuration eg : 
device_auhtorization_endpoint for google is fetched from this doc
https://accounts.google.com/.well-known/openid-configuration

* Adds the functionality to fetch the device code and user code along with verification URI from the auth server 
And  Prints this info in the console for the user to copy paste in the device he has access for using the browser 

```flytectl get projects --logger.level=6  --admin.authType DeviceFlow
{"json":{"src":"viper.go:398"},"level":"debug","msg":"Config section [storage] updated. No update handler registered.","ts":"2022-08-29T17:09:53+05:30"}
{"json":{"src":"viper.go:398"},"level":"debug","msg":"Config section [root] updated. No update handler registered.","ts":"2022-08-29T17:09:53+05:30"}
{"json":{"src":"viper.go:400"},"level":"debug","msg":"Config section [admin] updated. Firing updated event.","ts":"2022-08-29T17:09:53+05:30"}
{"json":{"src":"viper.go:398"},"level":"debug","msg":"Config section [files] updated. No update handler registered.","ts":"2022-08-29T17:09:53+05:30"}
{"json":{"src":"token_source_provider.go:253"},"level":"warning","msg":"Failed fetching from cache. Will restart the flow. Error: no token found in the cache","ts":"2022-08-29T17:09:54+05:30"}
Please open the browser at the url https://unionai.oktapreview.com/activate and enter following verification code HVTPGVTP
OR
Please open the browser at the url https://unionai.oktapreview.com/activate?user_code=HVTPGVTP containing verification code
Waiting for 10 secsWaiting for 20 secs{"json":{"src":"client.go:64"},"level":"info","msg":"Initialized Admin client","ts":"2022-08-29T17:10:27+05:30"}
{"json":{"src":"project.go:102"},"level":"debug","msg":"Retrieved 3 projects","ts":"2022-08-29T17:10:27+05:30"}
```

* Keeps polling the token endpoint until the user authorized  

```
 flytectl get projects --logger.level=6  --admin.authType DeviceFlow                                         
{"json":{"src":"viper.go:398"},"level":"debug","msg":"Config section [storage] updated. No update handler registered.","ts":"2022-08-29T17:40:27+05:30"}
{"json":{"src":"viper.go:398"},"level":"debug","msg":"Config section [root] updated. No update handler registered.","ts":"2022-08-29T17:40:27+05:30"}
{"json":{"src":"viper.go:400"},"level":"debug","msg":"Config section [admin] updated. Firing updated event.","ts":"2022-08-29T17:40:27+05:30"}
{"json":{"src":"viper.go:398"},"level":"debug","msg":"Config section [files] updated. No update handler registered.","ts":"2022-08-29T17:40:27+05:30"}
{"json":{"src":"token_source_provider.go:253"},"level":"warning","msg":"Failed fetching from cache. Will restart the flow. Error: no token found in the cache","ts":"2022-08-29T17:40:28+05:30"}
Please open the browser at the url https://unionai.oktapreview.com/activate and enter following verification code TNRJQDSD
OR
Please open the browser at the url https://unionai.oktapreview.com/activate?user_code=TNRJQDSD containing verification code
Waiting for 10 secs
Waiting for 20 secs
Waiting for 40 secs
Waiting for 80 secs

```

* Currently the wait time doubles with the minimum wait interval provided by authorization server
   If the server responds with slow_down or authentication_pending
* Final result of the fetch results with auth token  
```
flytectl get projects --logger.level=6  --admin.authType DeviceFlow                                         
{"json":{"src":"viper.go:398"},"level":"debug","msg":"Config section [storage] updated. No update handler registered.","ts":"2022-08-29T17:40:27+05:30"}
{"json":{"src":"viper.go:398"},"level":"debug","msg":"Config section [root] updated. No update handler registered.","ts":"2022-08-29T17:40:27+05:30"}
{"json":{"src":"viper.go:400"},"level":"debug","msg":"Config section [admin] updated. Firing updated event.","ts":"2022-08-29T17:40:27+05:30"}
{"json":{"src":"viper.go:398"},"level":"debug","msg":"Config section [files] updated. No update handler registered.","ts":"2022-08-29T17:40:27+05:30"}
{"json":{"src":"token_source_provider.go:253"},"level":"warning","msg":"Failed fetching from cache. Will restart the flow. Error: no token found in the cache","ts":"2022-08-29T17:40:28+05:30"}
Please open the browser at the url https://unionai.oktapreview.com/activate and enter following verification code TNRJQDSD
OR
Please open the browser at the url https://unionai.oktapreview.com/activate?user_code=TNRJQDSD containing verification code
Waiting for 10 secs
Waiting for 20 secs
Waiting for 40 secs
Waiting for 80 secs
{"json":{"src":"client.go:64"},"level":"info","msg":"Initialized Admin client","ts":"2022-08-29T17:43:01+05:30"}
{"json":{"src":"project.go:102"},"level":"debug","msg":"Retrieved 3 projects","ts":"2022-08-29T17:43:02+05:30"}
 --------------- --------------- --------------------------- 
| ID            | NAME          | DESCRIPTION               |
 --------------- --------------- --------------------------- 
| flyteexamples | flyteexamples | flyteexamples description |
 --------------- --------------- --------------------------- 
| flytetester   | flytetester   | flytetester description   |
 --------------- --------------- --------------------------- 
| flytesnacks   | flytesnacks   | flytesnacks description   |
 --------------- --------------- --------------------------- 
```

* Using caching of the token
```flytectl get projects --logger.level=6  --admin.authType DeviceFlow --admin.deviceFlowConfig.refreshTime 1m
{"json":{"src":"viper.go:398"},"level":"debug","msg":"Config section [storage] updated. No update handler registered.","ts":"2022-08-29T17:45:25+05:30"}
{"json":{"src":"token_orchestrator.go:87"},"level":"info","msg":"found the token in the cache","ts":"2022-08-29T17:45:25+05:30"}
{"json":{"src":"client.go:64"},"level":"info","msg":"Initialized Admin client","ts":"2022-08-29T17:45:25+05:30"}
{"json":{"src":"project.go:102"},"level":"debug","msg":"Retrieved 3 projects","ts":"2022-08-29T17:45:26+05:30"}
 --------------- --------------- --------------------------- 
| ID            | NAME          | DESCRIPTION               |
 --------------- --------------- --------------------------- 
| flyteexamples | flyteexamples | flyteexamples description |
 --------------- --------------- --------------------------- 
| flytetester   | flytetester   | flytetester description   |
 --------------- --------------- --------------------------- 
| flytesnacks   | flytesnacks   | flytesnacks description   |
 --------------- --------------- --------------------------- 
3 rows
```

* Using refresh token flow

```
flytectl get projects --logger.level=6  --admin.authType DeviceFlow --admin.deviceFlowConfig.refreshTime 10m

{"json":{"src":"token_orchestrator.go:60"},"level":"debug","msg":"got a response from the refresh grant for old expiry 2022-08-29 17:44:09.579357 +0530 IST with new expiry 2022-08-29 17:56:05.081999 +0530 IST m=+602.689286152","ts":"2022-08-29T17:46:05+05:30"}
{"json":{"src":"client.go:64"},"level":"info","msg":"Initialized Admin client","ts":"2022-08-29T17:46:05+05:30"}
{"json":{"src":"project.go:102"},"level":"debug","msg":"Retrieved 3 projects","ts":"2022-08-29T17:46:05+05:30"}
 --------------- --------------- --------------------------- 
| ID            | NAME          | DESCRIPTION               |
 --------------- --------------- --------------------------- 
| flyteexamples | flyteexamples | flyteexamples description |
 --------------- --------------- --------------------------- 
| flytetester   | flytetester   | flytetester description   |
 --------------- --------------- --------------------------- 
| flytesnacks   | flytesnacks   | flytesnacks description   |
 --------------- --------------- --------------------------- 
3 rows
```

* Max timeout for the entire workflow is defined using `timeout` flag which has default value of 10 mi
```
 --admin.deviceFlowConfig.timeout string   amount of time the device flow should complete or else it will be cancelled. (default "10m0s")
```

Pending unit tests and refactoring to utilize the same cache and refresh token flows across pkce and deviceflow 

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

 https://github.com/flyteorg/flyte/issues/2787

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by shahwar9** [comment]
This PR fixes this issue: https://github.com/flyteorg/flyte/issues/2787

**Review by shahwar9 — COMMENTED** [review]
Not sure if there is an auto documentation update, but probably updating `flytctl` documentation for the newly added SkipBrowserOpen is also required.

**Comment by pmahindrakar-oss** [comment]
@shahwar9 yes the docs would be updated automatically once this library is released and consumed in flytectl. 

eg : 

```
flytectl --help   
Flytectl is a CLI tool written in Go to interact with the FlyteAdmin service.

Usage:
  flytectl [command]
..........

      --admin.pkceConfig.refreshTime string        grace period from the token expiry after which it would refresh the token. (default "5m0s")
      --admin.pkceConfig.skipBrowserOpen           skips opening the browser for auth and instead print the url to use.
      --admin.pkceConfig.timeout string            Amount of time the browser session would be active for authentication from client app. (default "15s")
....
```

File: clients/go/admin/deviceflow/token_orchestrator.go [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+	if !token.Valid() {
+		return nil, fmt.Errorf("token from cache is invalid")
+	}
+
+	// If token doesn't need to be refreshed, return it.
+	if time.Now().Before(token.Expiry.Add(-t.cfg.TokenRefreshGracePeriod.Duration)) {
+		logger.Infof(ctx, "found the token in the cache")
+		return token, nil
+	}
+	token.Expiry = token.Expiry.Add(-t.cfg.TokenRefreshGracePeriod.Duration)
```
Comment (EngHabu) [inline clients/go/admin/deviceflow/token_orchestrator.go:90]: What is this needed for?

File: clients/go/admin/deviceflow/token_orchestrator.go [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+	httpResp, err := ctxhttp.Do(ctx, nil, httpReq)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := io.ReadAll(httpResp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("device authorization request failed due to  %v", err)
+	}
```
Comment (EngHabu) [inline clients/go/admin/deviceflow/token_orchestrator.go:126]: I think if StatusCode is not Ok, we shouldn't expect a body... should we check for StatusCode first?

File: clients/go/admin/deviceflow/token_orchestrator.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+		}
+	}
+
+	daresp := &DeviceAuthorizationResponse{}
+	err = json.Unmarshal(body, &daresp)
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Printf("Please open the browser at the url %v and enter following verification code %v\nOR\n", daresp.VerificationURL, daresp.UserCode)
```
Comment (EngHabu) [inline clients/go/admin/deviceflow/token_orchestrator.go:140]: The example you showed me with Google IdP actually has the user code as part of the url... is that not a supported mechanism by all IdPs?

File: clients/go/admin/deviceflow/token_orchestrator.go [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+// FetchTokenFromAuthFlow starts a webserver to listen to redirect callback from the authorization server at the end
+// of the flow. It then launches the browser to authenticate the user.
+func (t TokenOrchestrator) FetchTokenFromAuthFlow(ctx context.Context) (*oauth2.Token, error) {
+	ctx, cancelNow := context.WithTimeout(ctx, t.cfg.DeviceFlowTimeout.Duration)
+	defer cancelNow()
+
+	var scopes string
+	if len(t.clientConfig.Scopes) > 0 {
+		scopes = strings.Join(t.clientConfig.Scopes, " ")
+	}
```
Comment (EngHabu) [inline clients/go/admin/deviceflow/token_orchestrator.go:208]: If these are not specified, should we get the scopes from the Metadata endpoint?

File: clients/go/admin/deviceflow/token_orchestrator.go [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+	daReq := DeviceAuthorizationRequest{ClientID: t.clientConfig.ClientID, Scope: scopes}
+	daResp, err := t.StartDeviceAuthorization(ctx, daReq)
+	if err != nil {
+		return nil, err
+	}
+
+	pollInterval := int64(5) // default value of 5 sec poll interval if the authorization response doesn't have interval set
```
Comment (EngHabu) [inline clients/go/admin/deviceflow/token_orchestrator.go:215]: is it 5 seconds? can that move the config...

File: clients/go/admin/deviceflow/config.go [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package deviceflow
+
+import "github.com/flyteorg/flytestdlib/config"
+
+// Config defines settings used for Device orchestration flow.
+type Config struct {
+	TokenRefreshGracePeriod config.Duration `json:"refreshTime" pflag:",grace period from the token expiry after which it would refresh the token."`
+	DeviceFlowTimeout       config.Duration `json:"deviceFlowTimeout" pflag:",amount of time the device flow should complete or else it will be cancelled."`
```
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply clients/go/admin/deviceflow/config.go]: Done

File: clients/go/admin/deviceflow/token_orchestrator.go (hunk 2 above)
Comment (pmahindrakar-oss) [reply clients/go/admin/deviceflow/token_orchestrator.go:62]: The internal doTokenRoundTrip also compare the status code after reading the body 
https://github.com/golang/oauth2/blob/master/internal/token.go#L241

I think it should be fine since we throw an RetireveError which also send the body
```
&oauth2.RetrieveError{
			Response: httpResp,
			Body:     body,
		}
```

File: clients/go/admin/deviceflow/token_orchestrator.go (hunk 4 above)
Comment (pmahindrakar-oss) [reply clients/go/admin/deviceflow/token_orchestrator.go]: These are already got from the metadata endpoint. 
We can't send these from cli and is expected to be part of the flyteClientConfig which is returned by the metadata endpoint

File: clients/go/admin/deviceflow/token_orchestrator.go (hunk 5 above)
Comment (pmahindrakar-oss) [reply clients/go/admin/deviceflow/token_orchestrator.go]: Sure done

File: clients/go/admin/deviceflow/token_orchestrator.go (hunk 1 above)
Comment (pmahindrakar-oss) [reply clients/go/admin/deviceflow/token_orchestrator.go]: Currently there exists a bug in refresh token flow and even if the auth server return the refresh token , it never gets used . 

This is because , we use  reusableTokenSource https://github.com/golang/oauth2/blob/master/oauth2.go#L301
And which when figures out that it need to refresh the token but it can't since it has isValid check which compares the expiry.

And hence the token can't be refreshed until it expires which defeats the purpose of refresh token.

Hence we update the expiry with the grace period.

Added test to prove this issue

https://github.com/flyteorg/flyteidl/pull/313/files#diff-08990613781c77ceb555d931311713208a8d45ca6462eb52b87a98a452afe50d

File: clients/go/admin/deviceflow/config.go [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package deviceflow
+
+import "github.com/flyteorg/flytestdlib/config"
+
+// Config defines settings used for Device orchestration flow.
+type Config struct {
+	TokenRefreshGracePeriod config.Duration `json:"refreshTime" pflag:",grace period from the token expiry after which it would refresh the token."`
+	Timeout                 config.Duration `json:"deviceFlowTimeout" pflag:",amount of time the device flow should complete or else it will be cancelled."`
```
Comment (EngHabu) [inline clients/go/admin/deviceflow/config.go:8]:
```suggestion
	Timeout                 config.Duration `json:"timeout" pflag:",amount of time the device flow should complete or else it will be cancelled."`
```

File: clients/go/admin/deviceflow/token_orchestrator.go (hunk 2 above)
Comment (EngHabu) [reply clients/go/admin/deviceflow/token_orchestrator.go:62]: interesting... can we read it the same way then?
https://github.com/golang/oauth2/blob/master/internal/token.go#L236-L237
```
	body, err := ioutil.ReadAll(io.LimitReader(r.Body, 1<<20))
	r.Body.Close()
```

**Review by EngHabu — COMMENTED** [review]
LGTM just a couple of comments

File: clients/go/admin/deviceflow/token_orchestrator.go (hunk 2 above)
Comment (pmahindrakar-oss) [reply clients/go/admin/deviceflow/token_orchestrator.go]: Done

**Review by EngHabu — APPROVED** [review]

File: clients/go/admin/deviceflow/token_orchestrator.go [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+			return nil, err
+		}
+		httpReq.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+		httpResp, err := ctxhttp.Do(ctx, nil, httpReq)
+		if err != nil {
+			return nil, err
+		}
+
+		body, err := io.ReadAll(io.LimitReader(httpResp.Body, 1<<20))
```
Comment (katrogan) [inline clients/go/admin/deviceflow/token_orchestrator.go:107]: why the limit?

**Review by katrogan — APPROVED** [review]

File: clients/go/admin/deviceflow/token_orchestrator.go (hunk 8 above)
Comment (katrogan) [reply clients/go/admin/deviceflow/token_orchestrator.go:107]: resolved offline :)
